ballerina/engine.bal

```diff
@@ -21,20 +21,40 @@ import graphql.parser;
 isolated class Engine {
     private final readonly & __Schema schema;
     private final int? maxQueryDepth;
+    private final readonly & (readonly & Interceptor)[] interceptors;
+    private int nextInterceptor;
 
-    isolated function init(string schemaString, int? maxQueryDepth, Service s) returns Error? {
+    isolated function init(string schemaString, int? maxQueryDepth, Service s, (readonly & Interceptor)[] interceptors) returns Error? {
         if maxQueryDepth is int && maxQueryDepth < 1 {
             return error Error("Max query depth value must be a positive integer");
         }
         self.maxQueryDepth = maxQueryDepth;
         self.schema = check createSchema(schemaString);
+        self.interceptors = interceptors.cloneReadOnly();
+        self.nextInterceptor = 0;
         self.addService(s);
     }
 
     isolated function getSchema() returns readonly & __Schema {
         return self.schema;
     }
 
+    isolated function getInterceptors() returns (readonly & Interceptor)[] {
+        return self.interceptors;
+    }
+
+    isolated function getNextInterceptor() returns (readonly & Interceptor)? {
+        lock {
+            if self.interceptors.length() > self.nextInterceptor {
+                readonly & Interceptor next = self.interceptors[self.nextInterceptor];
+                self.nextInterceptor += 1;
+                return next;
+            }
+            self.nextInterceptor = 0;
```

Comment:
How does this work parallelly? If there are two requests processed parallelly, and one is processed the first interceptor, while the other has processed the third interceptor, what will be the result of this function?

The engine should not be used to store request-related data. The context should be used for that.